### DISCUSSION TITLE
howto build a relocatable bottle?

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2303] [Everyday usage]
been working with homebrew-freecad tap, and generating bottles using the provided brew test-bot. i've been able to generate some bottles using the CI/CD which makes it easier for install freecad dependent libraries, but however even generating a bottle, and then attempting to install such formula that has a bottle generated for it still requires building _such_ formula (opencascade 👀) from source. the reason it requires building opencascade from source is because the user i'm testing out these bottles has homebrew installed in a custom `brew --prefix` ie. `$USER/homebrew` i understand that's there's no official support for homebrew when homebrew is setup with a non `/usr/local` prefix, but something i've notice is that certain formula do not require installing _such_ formula from source and can be installed in my custom prefix without having to build from source whereas other formula (opencascade) require building from source.

how are certain formula able to be installed into a custom _brew prefix_ using the provided bottles without having to build the formula from source, whereas other formula require a rebuild?

i've looked out several different formula files and notice some have the below lines in the `bottle do` block, but don't quite understand how certain formula can be installed in a custom prefix without having to be rebuilt from source. 🤨

```
# bottle block from the current `curl.rb` formula
 bottle do
    sha256 cellar: :any,                 arm64_big_sur: "fb8bc11a92e6b9c69415ebcc0ac30a69d5fa7399ed48eda8f39acd08fb030f91"
    sha256 cellar: :any,                 big_sur:       "67eb98cd0d153e0e7ca5aadde8141da15cd6f47990136722e49e4a40b3d53c41"
    sha256 cellar: :any,                 catalina:      "c623592e8eb8cd9c4e60598f58e9e716b7daea398e75f207c23339624b8f4a71"
    sha256 cellar: :any,                 mojave:        "097d1eb078e53456d952030e0186eb153a515044336db22da0d7942eba21a7e0"
    sha256 cellar: :any_skip_relocation, x86_64_linux:  "730a10df5cd6086457c2ed329b0e881c6a8cd70590f5d7e5e731361438b822c3"
  end
```

noticing the `cellar: :any` ...but if i had homebrew setup on a linux box ie. linuxbrew since the above block has `cellar: :any_skip_relocation` does mean that the formula, _curl_ require building from source if linuxbrew is setup with a custom _brew prefix_?

my next question is, how can i make such formula, ie. (opencascade) relocatable to be installed in a custom _brew prefix_ without having to rebuild it? (on my current machines it almost takes an hour to rebuild) so  making subtle changes / tests to the formula file and then running it through the ci/cd pipeline becomes rather time consuming.

i assume that the upstream library / software needs to be able to support custom installation prefixes that can be relocated ie. being able to move such installation to a new directory after the `make` `make install` tasks have completed?

being able to identify how such formula, libraries, software that support such feature (relocation) would be rather handy. this all new to me, so forgive me if my questions / statements seem trivial.

**Comment by carlocab** [comment]
Bottles with `cellar: :any` or `cellar: :any_skip_relocation` can be installed into any prefix. The only non-relocatable bottles are the ones with no `cellar` parameter.

As for building relocatable bottles: it depends heavily on the build system. Making sure that there are no binaries that hardcode `HOMEBREW_PREFIX` is a good start, but I don't know of a general procedure that guarantees a bottle is relocatable. (If we did know of one, we'd probably already be doing it.)

Homebrew/brew#10846 might be a good place to start looking for ideas.

**Reply by HarikrishnanBalagopal** [reply]
Has support for catalina bottles been dropped? When I build bottles I only get Big Sur and Linux_x86_64 now

**Reply by carlocab** [reply]
This is due to your build setup. You can build bottles for any macOS version that you like (at least, anything newer than 10.10), but a bottle for a given macOS version needs to be built on that macOS version. You're getting Big Sur bottles because your build machine is running Big Sur.

**Reply by HarikrishnanBalagopal** [reply]
Is there a way to cross-compile for older or newer MacOS versions?
We are already using GoX to cross compile https://github.com/mitchellh/gox can we just give those binaries to brew to turn into bottles?
I am building bottles using Github actions, starting MacOS VMs is slow to begin with.
Having to start multiple MacOS versions just to build bottles is very slow and it's not great for billing either https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions#minute-multipliers

**Reply by SMillerDev** [reply]
If you have precompiled binaries that are exactly how you want them you might as well make it a cask.

**Reply by HarikrishnanBalagopal** [reply]
Isn't cask for GUI apps? Our project is a CLI tool.
Is it possible to do this with bottles instead of casks?

**Reply by ipatch** [reply]
@HarikrishnanBalagopal if you're worried about billing using github actions. i'd recommend using self-hosted runners. i have three macos vm's running on one macbook pro that work great using qemu+kvm as self hosted runners so i can build bottles for 10.14 10.15 and 10.16

**Reply by SMillerDev** [reply]
> Isn't cask for GUI apps? Our project is a CLI tool.

Casks are for precompiled binaries, and homebrew-casks requires them to be GUI or closed source. But for a personal tap you can do whatever you want.

**Reply by HarikrishnanBalagopal** [reply]
@ipatch Since ours is an open source project we can't use any private hosts or runners.

@SMillerDev
```
Casks are for precompiled binaries, and homebrew-casks requires them to be GUI or closed source. But for a personal tap you can do whatever you want.
```
oh ok. Ours is an open source project https://github.com/konveyor/homebrew-move2kube

To stick with the conventions that brew is using would it be possible to create bottles instead of casks (using the cross compiled binaries we generate) ?

**Reply by SMillerDev** [reply]
The convention for brew is that a bottle is compiled from source and a cask is precompiled. The rule I stated is just a rule for the official taps to avoid duplicate software.

You'd really be making everything more difficult for yourself if you deviate from that but in theory you can just make a formula that puts your precompiled file where you want it.

**Reply by ipatch** [reply]
@HarikrishnanBalagopal a `self-hosted` runner doesn't mean it has to be a "private runner". it just means you're setting up and maintaining the runner yourself. several of the homebrew projects use `self-hosted` runners. and i'm using `self-hosted` runners for the homebrew-freecad tap and freecad is GPL software. they allow a greater level of control as opposed to using the github provided runners.

**Reply by HarikrishnanBalagopal** [reply]
Thanks

**Comment by MikeMcQuaid** [comment]
Running `brew bottle -v` (even if you've already installed from a bottle) will help you by outputting what the hardcoded references are.
